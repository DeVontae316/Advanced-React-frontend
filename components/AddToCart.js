import React,{Component} from 'react';
import {Mutation} from 'react-apollo';
import gql from 'graphql-tag';

const ADD_TO_CART = gql`
 mutation ADD_TO_CART($id:ID!){
     addToCart(id:$id){
         id
         quantity 
     }
 }
`

class AddToCart extends Component{
    render(){
        const{id} = this.props;
        return (<Mutation  mutation={ADD_TO_CART}variables={{id:id}}> 
              {(addToCart,{loading,error})=>(
                  <button onClick={addToCart}>Add Item To Cart</button>
              )}
            </Mutation>
        )
        
    }
}


export default AddToCart;
export{ADD_TO_CART};