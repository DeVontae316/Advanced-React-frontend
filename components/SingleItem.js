import React,{Component}from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import{Query} from 'react-apollo';
import ItemStyles from './styles/ItemStyles';
import ErrorMessage from './ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head';
import PaginationStyles from './styles/PaginationStyles';

const Item = styled.div`
 max-width:1200px;
 margin: 2rem auto;
 box-shadow:${props => props.theme.bs};
 display:grid;
 grid-auto-colums: 1fr;
 grid-auto-flow:column;
 min-height:800px;
 img{
   width:100%;
   height:100%
   object-fit:contain;
 }
 .details{
   margin: 3rem;
   font-size: 2rem;

 }

`

const Center = styled.div`
 text-align:center;
`;

const SINGLE_CREATED_ITEM = gql`
 query SINGLE_CREATED_ITEM($id:ID!){
   item(where:{id:$id}){
     description
     title
     largeImage
     price
   }
 }
`

class SingleItem extends Component{

  render(){
    return(
    <Query query={SINGLE_CREATED_ITEM} variables={{id:this.props.id}}>
     {({data,loading,error})=>{
       if(loading) return <p>...loading</p>
       if(error) return <ErrorMessage error={error} />
       return<Item>
        <Head>
         <title>Sick Fits | {data.item.title}</title>
        </Head>
        <img src={data.item.largeImage} alt=" " />
        <div className="details">
         <h2>{data.item.title}</h2>
         <p>{data.item.description}</p>
        </div>
        {this.props.children}
       </Item>
       
     }}
    </Query>

    )
  }
}

export{SINGLE_CREATED_ITEM};
export default SingleItem;
