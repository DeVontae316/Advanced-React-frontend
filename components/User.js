import gql from 'graphql-tag';
import {Query} from 'react-apollo';
import PropTypes from 'prop-types';

const CURRENT_USER_QUERY = gql`
 query CURRENT_USER_QUERY{
   me{
     id
     email
     name
     permissions
    
     cart{
       id 
       quantity
       item{
         id
         image
         price 
         description
         title
       }
     }
   }
 }
`

const User = (props)=>(
  <Query {...props}query={CURRENT_USER_QUERY}>
   {payload => props.children(payload) }
  </Query>
)

User.propTypes = {
  children:PropTypes.func.isRequired,
}

export default User;
export {CURRENT_USER_QUERY};
