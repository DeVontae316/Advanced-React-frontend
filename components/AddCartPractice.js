import React from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import {Mutation} from 'react-apollo';
import Error from './ErrorMessage';
import {CURRENT_USER_QUERY} from './User';

const ADD_TO_CART =  gql`
mutation ADD_TO_CART($id:ID!){
    addToCart(id:$id){
        id 
        quantity
       
        
    }
}
`


const AddCartPractice = ({item})=>{
    return(
        <Mutation 
       
       refetchQueries={[{query:CURRENT_USER_QUERY}]}
        mutation={ADD_TO_CART} 
        variables={{id:item}}>
            {(addToCart,{loading,error})=>{
            
            if(error) return <Error error={error} />
             return <button disabled={loading} onClick={addToCart}>
                 Add To Cart Practice
             </button>

             }}
        </Mutation>
       
    )
}


export default AddCartPractice;
export{ADD_TO_CART}

