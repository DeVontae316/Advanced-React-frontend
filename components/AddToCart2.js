import React from 'react';
import gql from 'graphql-tag';
import{Mutation} from 'react-apollo';
import PropTypes from 'prop-types';
import AddToCart from './AddToCart';
import {CURRENT_USER_QUERY} from './User';

const ADD_TO_CART = gql`
 mutation ADD_TO_CART($id:ID!){
     addToCart(id:$id){
         id 
         quantity
         user{
             id 
             name
         }
     }
 }
`

const AddToCart2 = (props)=>{
    return(
        <Mutation refetchQueries={[{query:CURRENT_USER_QUERY}]} mutation={ADD_TO_CART} variables={{id:props.id}}>
           {(addToCart,{loading,error})=>(
               <button disabled={loading} onClick={addToCart}>ADD{loading && 'ing'} ITEM</button>
           )}
        </Mutation>
    )
}


export default AddToCart2;