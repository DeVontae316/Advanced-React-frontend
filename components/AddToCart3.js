import React from 'react';''
import{Mutation} from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import {CURRENT_USER_QUERY} from './User';

const ADD_ITEM_MUTATION = gql`
 mutation ADD_ITEM_MUTATION($id:ID!){
     addToCart(id:$id){
         id 
         quantity 
         user{
             id 
             email 
             name
         }
     }
 }
`;

const AddToCart3 = (props)=>{
    return(
       <Mutation refetchQueries={[{query:CURRENT_USER_QUERY}]} variables={{id:props.id}}mutation={ADD_ITEM_MUTATION}>
           {(addToCart,{loading,error})=>(
               <button onClick={addToCart}>ADD TO CART</button>
           )}
       </Mutation> 
        
    )
}


export  default AddToCart3;