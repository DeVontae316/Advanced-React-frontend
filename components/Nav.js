import Link from 'next/link';
import NavStyles from '../components/styles/NavStyles';
import User from './User';
import SignOut from './SignOut';
import {Mutation} from 'react-apollo';
import {LOCAL_STATE_MUTATION} from '../components/Cart';
import CartCount from './CartCount';
import styled from 'styled-components';
import Hooks from './Hooks';

const Cart = styled.p`
background:${props => props.theme.red};
color:white;
text-align:center;
padding:1.3rem;
line-height:2rem;
min-width:3rem;
margin-right:3rem;
font-weight:100;
font-feature-setting:'tnum';
font-variant-numeric:tabular-numeric;

`

const Nav = (props)=>(


    <User>
     {({data:{me}})=>(
     
      <NavStyles data-test="nav">

       <Link href='/items'>
        <a>Shop</a>

       </Link>
       {me &&(
        <>

         <Link href="/sell">
           <a>Sell</a>
         </Link>

         <Link href="/orders">
           <a>Orders</a>
         </Link>
         <Link href="test">
           <a>Account</a>
         </Link>
         
          <SignOut />
           {me.name}
          <Mutation mutation={LOCAL_STATE_MUTATION}>
            {(toggle,{loading,error})=>(
              

                <button onClick={toggle}><Cart>Cart</Cart>
                
                 <CartCount count={me.cart.reduce((reducer,item)=>{
                    const total = reducer + item.quantity;
                    return total;
                 },0)}/>
                 <p>items in cart</p>
                </button>
            )}
          </Mutation>
          
            
          
        </>

       )}
       {!me &&(
         <>
         <Link href='/signup'>
          <a>SignIn</a>
         </Link>
         </>
       )}


    </NavStyles>
     )}

    </User>

  )


export default Nav;
