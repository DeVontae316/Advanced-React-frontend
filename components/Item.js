import React,{Component} from 'react';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import Title from './styles/Title';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Link from 'next/link';
import formatMoney from '../lib/formatMoney';
import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';
import AddToCart2 from './AddToCart2';
import AddToCart3 from './AddToCart3';
import Modal from './Modal';
import AddCartPractice from './AddCartPractice';


class Item extends Component{
   static propTypes = {
     item:PropTypes.object.isRequired
   }
   state = {
     show:false
   }
   showModal = ()=>{
     this.setState({
       show:true,
     })
   }
  render(){
    const {item} = this.props;
    return(
      <ItemStyles>
       {item.image ? <img src={this.props.item.image}/> : null}
       <Title>
        <Link href={{pathname:"/item",query:{id:this.props.item.id}}}>
         <a>{this.props.item.title}</a>
        </Link>
       </Title>
       <PriceTag>{formatMoney(this.props.item.price)}</PriceTag>
       <p>{this.props.item.description}</p>
       <div className="buttonList">
       <Link href={{pathname:"/update",query:{id:item.id}}}>
        <a>update</a>
        </Link>
        
        <AddCartPractice item={item.id}/>
       <DeleteItem id={item.id}>Delete this item</DeleteItem>
        <button onClick={(e)=>{
          this.showModal();
        }}>Place Order</button>
        <Modal show={this.state.show}/>
       </div>
      </ItemStyles>
    )
  }
}


export default Item;
