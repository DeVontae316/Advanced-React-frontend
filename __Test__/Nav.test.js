import Nav from "../components/Nav";
import {mount} from 'enzyme';
import { EnzymeAdapter } from "enzyme";
import wait from 'waait';
import SingleItem,{SINGLE_CREATED_ITEM} from '../components/SingleItem';
import {MockedProvider} from 'react-apollo/test-utils';
import {fakeItem, fakeUser} from '../lib/testUtils';
import Item from "../components/styles/ItemStyles";
import toJSON from 'enzyme-to-json';
import {CURRENT_USER_QUERY} from "../components/User";

/*

What are the 3 steps behind creating creating mocks?

1.Retrieve query/mutation you desire.
2.Implement the query/mutation in your mock via the request property.
3.Set the response/data according to what you want to test.

After creating a MockProvider and providing it with mocks, what are 2 basic steps
you must implement if you're simulating a loading state?

1.await wait();
2.wrapper.update();

How would this apply to our SingleItem.test.js

1. Export the query from your SingleItem component
ie: export{Whatever name of query is}

2. const mock = [{
    request:{query:QUERY/MUTATION,variables:{id:"123"}
}]

3. In the example below, we are testing the query as if the id is valid and we expect a payload
const mock = [{
    request:{query:QUERY/MUTATION,variables:{id:"123"},
    result:{
        data:{
            item:fakeItem()
        }
    }
}]
3* The example below is as if we're expecting an error message:
 const mock = [{
    request:{query:QUERY/MUTATION,variables:{id:"123"},
    result:{
        erros:[{message:"Ooops!!!"}]
    }
}]

How would this apply to our Nav.test.js

1.Same as Single Item

2.const mocks = [
    {
        request:{query:YOUR_QUERY},
        
    }
]
3.The code below assumes the user "ISN'T" signed in
const notSignedIn = [
    {
        request:{query:YOUR_QUERY},
        result:{data:{me:null}}
        
    }
]
3. The code below assumes the user 'IS' signed in.
const SignedIn = [
    {
        request:{query:YOUR_QUERY},
        result:{data:{me:fakeUser()}}
        
    }
]
How do we implememt data-test attribute on a form?
1.Go to component and add data-test = "name"

How do we retrieve a element with the data-test attribute for snapshot testing?
1. const element = wrapper.find("[data-test]='nav'");
   expect(toJSON(element)).toMatchSnapshot();

*/

const notSignedIn = [{
    request:{query:CURRENT_USER_QUERY},
    result:{data:{
        me:null
    }},
    
}]

const SignedIn = [{
    request:{query:CURRENT_USER_QUERY},
    result:{data:{
        me:fakeUser()
    }},
    
}]

describe("<Nav/>",()=>{
    it("It renders a min nav when signed out!",async()=>{
      const wrapper = mount(
          <MockedProvider mocks={notSignedIn}>
              <Nav/>
          </MockedProvider>
      )
      await wait();
      wrapper.update();
      
      const nav = wrapper.find("[data-test='nav']");
      console.log(nav.debug());
      expect(toJSON(nav)).toMatchSnapshot();
    });
    it("It renders max nav when signed in!",async()=>{
        const wrapper = mount(
            <MockedProvider mocks={SignedIn}>
                <Nav/>
            </MockedProvider>
        )
        await wait();
        wrapper.update();
        
        const nav = wrapper.find("[data-test='nav']");
        const count = nav.find("div.count");
        console.log(nav.debug());
        expect(toJSON(count)).toMatchSnapshot();
      })

   
   
});