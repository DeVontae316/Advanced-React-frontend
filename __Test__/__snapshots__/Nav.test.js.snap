// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> It renders a min nav when signed out! 1`] = `
Array [
  <NavStyles
    data-test="nav"
  >
    <ul
      className="NavStyles-sc-11c0d2g-0 kHToJF"
      data-test="nav"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
        >
          Shop
        </a>
      </Link>
      <Link
        href="/signup"
      >
        <a
          href="/signup"
          onClick={[Function]}
        >
          SignIn
        </a>
      </Link>
    </ul>
  </NavStyles>,
  <ul
    className="NavStyles-sc-11c0d2g-0 kHToJF"
    data-test="nav"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
      >
        Shop
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        SignIn
      </a>
    </Link>
  </ul>,
]
`;

exports[`<Nav/> It renders max nav when signed in! 1`] = `
Array [
  <div
    className="count CartCount__Dot-xxvp4g-0 jqSlGR"
  >
    <div
      className="count"
    >
      0
    </div>
  </div>,
  <div
    className="count"
  >
    0
  </div>,
]
`;
